client/state/posts/reducer.js

```diff
@@ -47,12 +56,36 @@ export function items( state = {}, action ) {
 	switch ( action.type ) {
 		case POSTS_RECEIVE:
 			return Object.assign( {}, state, keyBy( action.posts, 'global_ID' ) );
+
+		case POST_DELETE:
+			const globalId = findKey( state, {
+				ID: action.postId,
+				site_ID: action.siteId
+			} );
+
+			if ( ! globalId ) {
+				break;
+			}
+
+			// Posts and pages are first sent to trash before being permanently
+			// deleted. Therefore, only omit if already trashed or custom type.
+			const post = state[ globalId ];
+			if ( 'trash' === post.status || ! includes( [ 'post', 'page' ], post.type ) ) {
```

Comment:
>This feels odd that we don't have a separate trash action. 

Indeed, this is unfortunately an artifact of the behavior of the REST API ([documentation](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/delete/))